server/src/main/java/org/elasticsearch/cluster/routing/allocation/ShardsAllocationHealthIndicatorService.java

```diff
@@ -0,0 +1,164 @@
+ * <p>
+ * Each shard needs to be available and replicated in order to guarantee high availability and prevent data loses.
+ */
+public class ShardsAllocationHealthIndicatorService implements HealthIndicatorService {
+
+    public static final String NAME = "shards_allocation";
+
+    private final ClusterService clusterService;
+
+    public ShardsAllocationHealthIndicatorService(ClusterService clusterService) {
+        this.clusterService = clusterService;
+    }
+
+    @Override
+    public String name() {
+        return NAME;
+    }
+
+    @Override
+    public String component() {
+        return DATA;
+    }
+
+    @Override
+    public HealthIndicatorResult calculate() {
+
+        var state = clusterService.state();
+        var status = new ShardAllocationStatus();
+
+        for (IndexRoutingTable indexShardRouting : state.routingTable()) {
+            for (IndexShardRoutingTable shardRouting : indexShardRouting) {
+                status.addPrimary(shardRouting.primaryShard());
+                for (ShardRouting replicaShard : shardRouting.replicaShards()) {
+                    status.addReplica(replicaShard);
+                }
+            }
+        }
+
+        return createIndicator(status.getStatus(), status.getSummary(), status.getDetails());
+    }
+
+    private static class ShardAllocationCounts {
+        private int unassigned = 0;
+        private int unassigned_restarting = 0;
+        private int initializing = 0;
+        private int started = 0;
+        private int relocating = 0;
+
+        public void increment(ShardRouting routing) {
+            switch (routing.state()) {
+                case UNASSIGNED -> {
+                    unassigned++;
+                    if (routing.unassignedInfo() != null && routing.unassignedInfo().getReason() == UnassignedInfo.Reason.NODE_RESTARTING) {
+                        unassigned_restarting++;
+                    }
+                }
+                case INITIALIZING -> initializing++;
+                case STARTED -> started++;
+                case RELOCATING -> relocating++;
+            }
+        }
+    }
+
+    private static class ShardAllocationStatus {
+        private final ShardAllocationCounts primaries = new ShardAllocationCounts();
+        private final ShardAllocationCounts replicas = new ShardAllocationCounts();
+
+        public void addPrimary(ShardRouting routing) {
+            primaries.increment(routing);
+        }
+
+        public void addReplica(ShardRouting routing) {
+            replicas.increment(routing);
+        }
+
+        public HealthStatus getStatus() {
+            if (primaries.unassigned > 0) {
+                return RED;
+            } else if (replicas.unassigned > replicas.unassigned_restarting) {
+                return YELLOW;
+            } else {
+                return GREEN;
+            }
+        }
+
+        public String getSummary() {
+            var builder = new StringBuilder("This cluster has ");
```

Comment:
I spoked with Henning about the restarting thing earlier and was convinced by his opinion of including it. I do think we need a special summary when there are unavailable shards due to restarts, but the status is still green.

"This cluster has 3 temporarily unavailable replicas due to a node restarting." Or something like that. If the status is yellow or red I think we can just ignore the restarting terminology and reference the unavailable shards.